azkaban-common/src/main/java/azkaban/container/models/AzKubernetesV1PodTemplate.java

```diff
@@ -0,0 +1,140 @@
+ * AzKubernetesV1PodTemplate)} with the pod-spec created using {@link AzKubernetesV1SpecBuilder}
+ * <p>
+ * Merging criterion is such that, if the item in the already created pod-spec has the same name as
+ * of the item extracted from the template, then it will be retained.
+ * <p>
+ * This template design enables declaration of static initContainers, volumes, volumeMounts, etc.
+ * which are not job-types but are useful. For example: init container which will fetch the required
+ * certificates/tokens etc. writes to a volume and then that volume will be mounted to the
+ * application-container/flow-container.
+ */
+public class AzKubernetesV1PodTemplate {
+
+  private V1Pod podFromTemplate;
+  private static AzKubernetesV1PodTemplate instance;
+
+  /**
+   * Private constructor to make this class singleton
+   *
+   * @param templatePath th where the template file is located.
+   * @throws IOException If unable to read the template file.
+   */
+  private AzKubernetesV1PodTemplate(String templatePath) throws IOException {
+    File templateFile = Paths.get(templatePath).toFile();
+    this.podFromTemplate = (V1Pod) Yaml.load(templateFile);
+  }
+
+  /**
+   * @param templatePath Path where the template file is located.
+   * @return Singleton instance of this class.
+   * @throws IOException If unable to read the template file.
+   */
+  public static synchronized AzKubernetesV1PodTemplate getInstance(String templatePath)
+      throws IOException {
+    if (null == instance) {
+      instance = new AzKubernetesV1PodTemplate(templatePath);
+    }
+    return instance;
+  }
+
+  /**
+   * @return the {@link V1Pod} POD generated from the template.
+   */
+  public V1Pod getPodFromTemplate() {
+    return this.podFromTemplate;
+  }
+
+  /**
+   * @param filterPredicate Predicate to filter the init containers.
+   * @return The list of filtered init Containers derived from the POD specified in the template.
+   */
+  public List<V1Container> getInitContainers(Predicate<? super V1Container> filterPredicate) {
+    V1PodSpec spec = this.podFromTemplate.getSpec();
+    if (null == spec) {
+      return Collections.emptyList();
+    }
+    List<V1Container> initContainers = spec.getInitContainers();
+    return initContainers == null ? Collections.emptyList() :
+        initContainers.stream().filter(filterPredicate).collect(Collectors.toList());
+  }
+
+  /**
+   * @return The list of init Containers derived from the POD specified in the template.
+   */
+  public List<V1Container> getInitContainers() {
+    V1PodSpec spec = this.podFromTemplate.getSpec();
+    return spec == null ? Collections.emptyList() : spec.getInitContainers();
+  }
+
+  /**
+   * @param filterPredicate Predicate to filter the volumes.
+   * @return The list of filtered Volumes derived from the POD specified in the template.
+   */
+  public List<V1Volume> getVolumes(Predicate<? super V1Volume> filterPredicate) {
+    V1PodSpec spec = this.podFromTemplate.getSpec();
```

Comment:
We have the AzKubernetesV1PodTemplate as a singleton. When instantiated, podFromTemplate will always be non-null as we are using "Yaml.load()" which will not return null but will fail in case it is unable to read or even if the file is empty. Hence it will never be null.